concordium-std/src/bump_alloc.rs

```diff
@@ -0,0 +1,184 @@
+            *heap_end = actual_size;
+        }
+
+        // Align the address.
+        let alloc_start = align_up(*next, layout.align());
+        // Get the end of the allocation. This should always return `Some` as the
+        // contract memory is limited to much below usize::MAX.
+        let alloc_end = match alloc_start.checked_add(layout.size()) {
+            Some(end) => end,
+            None => return ptr::null_mut(),
+        };
+
+        // Check if we need to request more memory.
+        if alloc_end > *heap_end {
+            let space_needed = alloc_end - *heap_end;
+            let pages_to_request = PageCount(div_ceil(space_needed, PAGE_SIZE));
+            let previous_page_count = self.grow_memory(pages_to_request);
+            // Check if we are out of memory.
+            if previous_page_count == ERROR_PAGE_COUNT {
+                return ptr::null_mut();
+            }
+            // Increase the heap size.
+            *heap_end += pages_to_request.size_in_bytes();
+        }
+
+        // Increment allocations counter.
+        *self.allocations.get() += 1;
+        // Remember the last address handed out, so that we may move next backwards if
+        // it is deallocated before a new allocation occurs.
+        *self.last_alloc.get() = alloc_start;
+        *next = alloc_end;
+        alloc_start as *mut u8
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
+        let allocations = self.allocations.get();
+        let last_alloc = self.last_alloc.get();
+        let next = self.next.get();
+        // Decrease the allocation counter.
+        *allocations -= 1;
+        // Reset next and last allocation if everything has been deallocated.
+        if *allocations == 0 {
+            *next = MIN_PTR_ADDR;
+            *last_alloc = MIN_PTR_ADDR;
+        } else if *last_alloc as *mut u8 == ptr {
+            // Move next back to last alloc. This is a small optimization over the regular
+            // bump allocator.
+            *next = *last_alloc;
+        }
+    }
+}
+
+/// Align the given address, `addr`, upwards to alignment `align`.
+///
+/// Requires that `align` is a power of two, which it always is due to the
+/// design of [`Layout`].
+///
+/// Uses a bitmask to align the addresses for efficiency.
+/// For details, see: https://os.phil-opp.com/allocator-designs/#address-alignment
+fn align_up(addr: usize, align: usize) -> usize { (addr + align - 1) & !(align - 1) }
+
+/// Calculates the quotient of `lhs / rhs`, while rounding up to the nearest
+/// integer.
+///
+/// ```
+/// let a = 8;
+/// let b = 3;
+/// assert_eq!(a.div_ceil(b), 3);
+/// assert_eq!(a.div_ceil(-b), -2);
+/// assert_eq!((-a).div_ceil(b), -2);
+/// assert_eq!((-a).div_ceil(-b), 3);
+/// ```
+const fn div_ceil(lhs: usize, rhs: usize) -> usize {
+    let d = lhs / rhs;
+    let r = lhs % rhs;
+    if r > 0 {
+        d + 1
+    } else {
+        d
+    }
```

Comment:
**Suggestion:** I did not test this, but I think we could avoid branching here, maybe the compiler already does this already internally.
```suggestion
    let rounding = r > 0 as usize;
    d + rounding
```